### PR TITLE
Set timeout on all requests.get calls

### PR DESCRIPTION
**Pull request by Androbin**
See #343 for reference
The change in `web.py` will propagate to all modules using it.

Note that this only makes requests fail faster.
If we want to skip tests, we will have to do this in a follow-up.

### CODE REVIEW

**Comment by shardulc** [comment]
Looks good, although where does the magic number 61 come from?

**Comment by Androbin** [comment]
That's the number of seconds to wait. One minute + epsilon.
I suspected we would argue about the exact value anyway.
@shardulc Would a common global for this be appropriate?

**Comment by shardulc** [comment]
Common global in `web.py` would do, I think. Also, 61 seconds is a pretty long time to wait; does 5 seem reasonable?

**Comment by Androbin** [comment]
> 61 seconds is a pretty long time to wait

It was deliberately that high in the first draft since I didn't want to step on anyone's toes.

> does 5 seem reasonable?

We should not depend on too good of a connection. Even 10 seconds may not be that long.
If it's faster, good. If the connection is not fast, we don't want to fail after just 5 seconds.
On the other hand, failing after 10 seconds feels much better than eventually failing after half an hour.

Note that the changes don't just apply to our tests but to all executed requests.

**Comment by Androbin** [comment]
In the case of timeout, a `requests.exceptions.ReadTimeout` is raised.
The latter could be caught by the tests.

**Comment by shardulc** [comment]
I agree. I think some tests use an `is_up` function to check web APIs
before they use them, can you check the timeouts on those too?

**Comment by Androbin** [comment]
> I think some tests use an `is_up` function to check web APIs

The `is_up` function currently sets a timeout of 4 seconds and checks for the exception.

**Comment by Androbin** [comment]
> if there is a timeout, we should skip the test instead of failing at the timeout error, right?

How about something like https://stackoverflow.com/a/34659817 for convenience?

**Comment by shardulc** [comment]
That works. So, we can move the `is_up` code to `web.py` and turn it into a decorator that takes 1 argument (the web service to check) and skips the test if the service is down.

**Comment by Androbin** [comment]
> we can move the is_up code to web.py and turn it into a decorator

Not quite. There are exactly 2 uses outside the tests, namely in the `search` module.
But we can both move the function and create a decorator using it.

**Comment by Androbin** [comment]
@shardulc I ended up touching more lines than expected.
Hopefully, reviewing is not too much of a burden.
By the way, there is still no new task applicable to `begiak`.

File: web.py [hunk 1]
```diff
@@ -20,19 +31,32 @@ def http_error_default(self, url, fp, errcode, errmsg, headers):
...
 
+def is_up(url):
+    global up_down
 
-from requests.exceptions import ConnectionError, HTTPError, InvalidURL
-from html.entities import name2codepoint
-from urllib.parse import quote, unquote
+    if (url not in up_down) or (time() - up_down[url][1] > 600):
+        try:
+            requests.get(url, timeout=5).raise_for_status()
```
Comment (shardulc) [inline web.py]: Should use `REQUEST_TIMEOUT`

**Comment by shardulc** [comment]
I think there was a misunderstanding about the decorator. What I meant was that we would have a decorator  for the entire class, like
```python
@catchTimeout('https://something')
class TestSomething:
    etc.
```
which would not require changes to every single test.

File: modules/linx.py [hunk 2]
```diff
@@ -1,40 +1,40 @@
-#!/usr/bin/python3
```
[the remark this answers is not in the document]
Comment (Androbin) [reply modules/linx.py]: Some files apparently had Windows CRLF line encoding.
Changed it to Unix LF line encoding.

File: web.py (hunk 1 above)
Comment (Androbin) [reply web.py]: Really? Was 4 before, now 10?

File: modules/linx.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (shardulc) [reply modules/linx.py]: From a stylistic point of view, that's fine, but I would rather not have the huge changeset in this PR. Plus it makes git blame etc. harder.

File: web.py (hunk 1 above)
Comment (shardulc) [reply web.py]: "4 before": ? The whole reason we chose 10 was because you thought we shouldn't depend on very good connections?

**Comment by Androbin** [comment]
Does it make sense to pass `catch_timeout` all domain names we encounter?

File: web.py (hunk 1 above)
Comment (Androbin) [reply web.py]: Alright. Only the `search.py` module uses this anyway...

File: modules/linx.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Androbin) [reply modules/linx.py]: Okay, reverted.
Shall I send another PR?

**Comment by shardulc** [comment]
Yes, that was what I had in mind. Before we even run the test, we check all domains that could possibly be used. I was thinking that the class decorator `@catch_timeout` would use [this](https://docs.python.org/3/library/unittest.html#skipping-tests-and-expected-failures): "It’s easy to roll your own skipping decorators by making a decorator that calls skip() on the test when it wants it to be skipped". It is probably more elegant than calling another decorator internally on each function.

**Comment by Androbin** [comment]
Before, we had `is_up` tests in the `set_up` or `setUpClass` methods.
The motivation behind this change is to catch timeouts on the actual calls which may fail even if the previous check passed.
Don't forget that the return value of `is_up` is cached and that the actual value can differ even within a second.

**Comment by shardulc** [comment]
Merging this. The begiak task should be available now.

**Comment by Androbin** [comment]
Just realized we should be able to `socket.setdefaulttimeout(REQUEST_TIMEOUT)`.

**Comment by shardulc** [comment]
Is `socket.setdefaulttimeout` required in addition to the timeouts we've already set, or is it one or the other?

**Comment by Androbin** [comment]
We already set timeouts on `requests.get`, directly or via `web.py`.
But e.g. `clock.py > npl` uses raw sockets which may get stuck on connect.
With #379 we could set timeouts for all sockets and fix e.g. [this Travis build](https://travis-ci.org/goavki/phenny/builds/329949041).

**Comment by Androbin** [comment]
@shardulc What you think?

**Comment by Androbin** [comment]
With #379 we could set timeouts for all sockets and fix e.g. [this Travis build](https://travis-ci.org/goavki/phenny/builds/329949041).

**Pull request merged**